pkg/solana/txm/txm.go

```diff
@@ -303,65 +295,224 @@ func (txm *Txm) sendWithRetry(ctx context.Context, msg pendingTx) (solanaGo.Tran
-						}
-						txm.lggr.Debugw("tx rebroadcast with bumped fee", "id", msg.id, "fee", getFee(count), "signatures", sigs.List())
-					}
+	// return signed tx, id, signature for use in simulation
+	return initTx, msg.id, sig, nil
+}
 
-					// prevent locking on waitgroup when ctx is closed
-					wait := make(chan struct{})
-					go func() {
-						defer close(wait)
-						sigs.Wait(count) // wait until bump tx has set the tx signature to compare rebroadcast signatures
-					}()
-					select {
-					case <-ctx.Done():
-						return
-					case <-wait:
-					}
+// buildTx constructs a new transaction with the specified fee and signs it.
+// It updates the compute unit price and appends the signature to the transaction.
+func (txm *Txm) buildTx(ctx context.Context, baseTx solanaGo.Transaction, key string, retryCount int) (solanaGo.Transaction, error) {
+	// Set fee
+	if computeUnitErr := fees.SetComputeUnitPrice(&baseTx, txm.getFee(retryCount)); computeUnitErr != nil {
+		return solanaGo.Transaction{}, computeUnitErr
+	}
 
-					// this should never happen (should match the signature saved to sigs)
-					if fetchedSig, fetchErr := sigs.Get(count); fetchErr != nil || retrySig != fetchedSig {
-						txm.lggr.Errorw("original signature does not match retry signature", "expectedSignatures", sigs.List(), "receivedSignature", retrySig, "error", fetchErr)
-					}
-				}(shouldBump, bumpCount, currentTx)
+	// Serialize the transaction message for signing
+	txMsg, marshalErr := baseTx.Message.MarshalBinary()
+	if marshalErr != nil {
+		return solanaGo.Transaction{}, fmt.Errorf("error in MarshalBinary: %w", marshalErr)
+	}
+
+	// Sign the transaction message using the key
+	sigBytes, signErr := txm.ks.Sign(ctx, key, txMsg)
+	if signErr != nil {
+		return solanaGo.Transaction{}, fmt.Errorf("error in Sign: %w", signErr)
+	}
+
+	// Append the signature to the transaction
+	var finalSig [64]byte
+	copy(finalSig[:], sigBytes)
+	baseTx.Signatures = append(baseTx.Signatures, finalSig)
+
+	return baseTx, nil
+}
+
+// getFee calculates the compute unit price for a transaction based on the number of retries.
+func (txm *Txm) getFee(count int) fees.ComputeUnitPrice {
+	// base compute unit price should only be calculated once
+	// prevent underlying base changing when bumping (could occur with RPC based estimation)
+	fee := fees.CalculateFee(
+		txm.cfg.ComputeUnitPriceDefault(),
+		txm.cfg.ComputeUnitPriceMax(),
+		txm.cfg.ComputeUnitPriceMin(),
+		uint(count), //nolint:gosec // reasonable number of bumps should never cause overflow
+	)
+	return fees.ComputeUnitPrice(fee)
+}
+
+// bumpFeeAndUpdateBlockhash determines whether the transaction fee should be bumped
+// and whether the blockhash needs to be updated based on the configured criteria.
+func (txm *Txm) bumpFeeAndUpdateBlockhash(ctx context.Context, txID string, bumpTime time.Time) (shouldBump bool, needBlockhashUpdate bool, newBumpTime time.Time, err error) {
```

Comment:
I noticed we call this in the retry routine for each transaction. I think that could lead to more RPC calls than we need to make for this. Otherwise, we'd be fetching `SlotHeight` for each transaction that's reached its retry interval. I feel like we could use a TXM level go routine like the `reap` or `confirm` loops to handle expiration. That way we only call `SlotHeight` once every time we reach some "check expiration" interval then evaluate all transactions in `Broadcasted` state against it